### PR TITLE
[FLINK-13749][client] Make PackagedProgram respect classloading policy

### PR DESCRIPTION
**Pull request by link3280**
## What is the purpose of the change
Currently, Flink client (PackagedProgram) does not respect the classloading policy and uses hardcoded parent-first classloader, while the other components like jobmanager and taskmanager use child-first classloader by default and respect the classloading options. This makes the client more likely to have dependency conflicts, especially after we removed the convenient hadoop binaries (so users need to add hadoop classpath in the client classpath).

## Brief change log

  - *Add configuration to PackagedProgram construtors.*
  - *Determind PackagedProgram classloader with respect to classloader resolve-order and parent-first pattern in the configuration.*

## Verifying this change

This change added tests and can be verified as follows:

- Add tests in ClassLoaderITCase to check classloading orders in PackagedProgram execution.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by link3280** [comment]
cc @TisonKun

**Comment by zentol** [comment]
cc @NicoK

File: flink-tests/src/test/java/org/apache/flink/test/classloading/jar/ClassLoadingPolicyProgram.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+public class ClassLoadingPolicyProgram {
+
+	public static void main(String[] args) throws Exception {
+		if (args.length < 2) {
+			throw new IllegalArgumentException("Missing parameters");
+		}
+		String resourceName = args[0];
+		String expectedResourceDir = args[1];
+		URL url = Thread.currentThread().getContextClassLoader().getResource(resourceName);
+		if (url == null) {
```
Comment (TisonKun) [inline flink-tests/src/test/java/org/apache/flink/test/classloading/jar/ClassLoadingPolicyProgram.java]: `checkNotNull`

**Comment by aljoscha** [comment]
@TisonKun there is this one: https://github.com/apache/flink/pull/9683

**Comment by link3280** [comment]
Thanks for the comments. The PR is updated.

File: flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java [hunk 2]
```diff
@@ -91,15 +93,21 @@ public static void checkJarFile(URL jar) throws IOException {
 		}
 	}
 
-	public static ClassLoader buildUserCodeClassLoader(List<URL> jars, List<URL> classpaths, ClassLoader parent) {
+	public static ClassLoader buildUserCodeClassLoader(List<URL> jars, List<URL> classpaths, ClassLoader parent,
+														Configuration configuration) {
```
Comment (NicoK) [inline flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java]: please use one parameter per line

**Comment by link3280** [comment]
Thanks a lot for your detailed review @NicoK ! The comments are addressed.

Agreed with the `configuration` issue. It's more user-friendly and concise to pass `configuration` instead of the concrete options.

Regarding the regression problem, I'm not sure I fully understand the situation. Does the old code mean a previous commit in this PR? If yes, it should be fixed already.

**Comment by aljoscha** [comment]
Thanks, I merged this!

**Comment by aljoscha** [comment]
@link3280 could you maybe also open PRs against the `release-1.8` and `release-1.9` branches, so that we also have the fix available when we release 1.8.x and 1.9.x?

**Comment by link3280** [comment]
@aljoscha No problem :)

**Comment by hequn8128** [comment]
@link3280 I'm checking the release note of 1.9.2 and found that the fix has not been included in 1.9, could you also open PRs against release-1.9 as @aljoscha suggested?
BTW, you do not need to open PRs against 1.8 anymore since we are releasing 1.10 currently.

**Comment by link3280** [comment]
@hequn8128 Thanks a lot for your work on the new release! But IIUC, I did have backported this PR to release-1.9, and here's the PR https://github.com/apache/flink/pull/10243 the commit https://github.com/apache/flink/commit/0f30c263eebd2fc3ecbeae69a4ce9477e1d5d774 .

**Comment by hequn8128** [comment]
@link3280 Thanks, I see. I failed to find the PR as the commit log only include `[FLINK-13747]`. Should it be `[FLINK-13749]`?

**Comment by link3280** [comment]
@hequn8128 Yes. Sorry for the mistake and the inconvenience it caused.

**Comment by hequn8128** [comment]
@link3280 No worry, thanks a lot for your contribution!

**Pull request closed without merging**